crates/bevy_ecs/src/schedule_v3/config.rs

```diff
@@ -0,0 +1,618 @@
+use bevy_utils::prelude::default;
+use bevy_utils::HashSet;
+
+use crate::{
+    schedule_v3::{
+        condition::{BoxedCondition, Condition},
+        graph::{Ambiguity, DependencyEdgeKind, GraphInfo},
+        set::{BoxedSystemSet, IntoSystemSet, SystemSet},
+    },
+    system::{BoxedSystem, IntoSystem, System},
+};
+
+/// A [`SystemSet`] with scheduling metadata.
+pub struct SystemSetConfig {
+    pub(super) set: BoxedSystemSet,
+    pub(super) graph_info: GraphInfo,
+    pub(super) conditions: Vec<BoxedCondition>,
+}
+
+/// A [`System`] with scheduling metadata.
+pub struct SystemConfig {
+    pub(super) system: BoxedSystem,
+    pub(super) graph_info: GraphInfo,
+    pub(super) conditions: Vec<BoxedCondition>,
+}
+
+pub(super) fn new_set_unchecked(set: BoxedSystemSet) -> SystemSetConfig {
+    SystemSetConfig {
+        set,
+        graph_info: GraphInfo {
+            sets: HashSet::new(),
+            dependencies: HashSet::new(),
+            ambiguous_with: default(),
+        },
+        conditions: Vec::new(),
+    }
+}
+
+fn new_set(set: BoxedSystemSet) -> SystemSetConfig {
+    assert!(!set.is_system_type());
+    new_set_unchecked(set)
+}
+
+fn new_system(system: BoxedSystem) -> SystemConfig {
+    // include system in its default sets
+    let sets = system.default_system_sets().into_iter().collect();
+    SystemConfig {
+        system,
+        graph_info: GraphInfo {
+            sets,
+            dependencies: HashSet::new(),
+            ambiguous_with: default(),
+        },
+        conditions: Vec::new(),
+    }
+}
+
+fn new_condition<P>(condition: impl Condition<P>) -> BoxedCondition {
+    let condition_system = IntoSystem::into_system(condition);
+    assert!(
+        condition_system.is_send(),
+        "condition accesses thread-local resources (currently not supported)"
+    );
+
+    Box::new(condition_system)
+}
+
+/// Types that can be converted into a [`SystemSetConfig`].
+///
+/// This has been implemented for all types that implement [`SystemSet`] and boxed trait objects.
+pub trait IntoSystemSetConfig: sealed::IntoSystemSetConfig {
+    /// Convert into a [`SystemSetConfig`].
+    #[doc(hidden)]
+    fn into_config(self) -> SystemSetConfig;
+    /// Add to `set` membership.
+    fn in_set(self, set: impl SystemSet) -> SystemSetConfig;
+    /// Run before all members of `set`.
+    fn before<M>(self, set: impl IntoSystemSet<M>) -> SystemSetConfig;
+    /// Run after all members of `set`.
+    fn after<M>(self, set: impl IntoSystemSet<M>) -> SystemSetConfig;
+    /// Run only if the [`Condition`] is `true` at the time of execution.
```

Comment:
We should be more precise about timing details in the docs. Probably in `Condition`?